kbfsgit/runner.go

```diff
@@ -1423,26 +1430,81 @@ func (r *runner) pushAll(ctx context.Context, fs *libfs.FS) (err error) {
 		fmt.Sprintf("Preparing and %s packed refs", verb), "pushprefs")
 }
 
+// parentCommitsForRef returns a map of refs with a list of commits for each
+// ref, newest first. It only includes commits that exist in `localStorer` but
+// not in `remoteStorer`.
+func (r *runner) parentCommitsForRef(ctx context.Context,
+	localStorer gogitstor.Storer, remoteStorer gogitstor.Storer,
+	refs refLookupByName) (commitsByRefName, error) {
+
+	commitsByRef := make(commitsByRefName, len(refs))
+	haves := make(map[plumbing.Hash]bool)
+
+	for _, ref := range refs {
+		hash := ref.Hash()
+		refName := ref.Name()
+
+		// Get the HEAD commit for the ref from the local repository.
+		commit, err := gogitobj.GetCommit(localStorer, hash)
+		if err != nil {
+			r.log.CDebugf(ctx, "Error getting commit for hash %s: %+v",
+				string(hash[:]), err)
+			continue
+		}
+
+		// Iterate through the commits backward, until we experience any of the
+		// following:
+		// 1. Find a commit that the remote knows about,
+		// 2. Reach our maximum number of commits to check,
+		// 3. Run out of commits.
+		walker := gogitobj.NewCommitPreorderIter(commit, haves, nil)
+		toVisit := maxCommitsToVisitPerRef
+		err = walker.ForEach(func(c *gogitobj.Commit) error {
+			haves[c.Hash] = true
+			toVisit--
+			// If toVisit starts out at 0 (indicating there is no
+			// max), then it will be negative here and we won't stop
+			// early.
+			hasEncodedObjectErr := remoteStorer.HasEncodedObject(c.Hash)
+			if toVisit == 0 || hasEncodedObjectErr == nil {
```

Comment:
I slightly prefer `if toVisit == 0 || remoteStorer.HasEncodedObject(c.Hash) == nil` here but I'm OK with this too.

Reply:
I had it like that first and found myself having to mentally detangle the cases...separated out it seemed easier to reason through.

Reply:
The issue is that `HasEncodedObject` is terribly named: a name like that should return a `bool`, and it's confusing that it doesn't.

Reply:
Oh I see. 

> a name like that should return a bool

Yeah agree; that was why I didn't realize this was what deals with the case where a commit is found in remote branch, and my motivation of making this comment.

But the three points above are helpful enough so carry on!